doc/tutorials/your_first_project/your_first_project.rst

```diff
@@ -0,0 +1,275 @@
+
+2 Create a ROS Node and Executor
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This first bit is a bit of boilerplate but you should be used to seeing this from the ROS 2 tutorials.
+
+.. code-block:: C++
+
+  #include <memory>
+  #include <thread>
+
+  #include <rclcpp/rclcpp.hpp>
+  #include <moveit/move_group_interface/move_group_interface.h>
+
+  int main(int argc, char * argv[])
+  {
+    // Initialize ROS and create the Node
+    rclcpp::init(argc, argv);
+    auto const node = std::make_shared<rclcpp::Node>(
+      "hello_moveit",
+      rclcpp::NodeOptions().automatically_declare_parameters_from_overrides(true)
+    );
+
+    // Create a ROS logger
+    auto const logger = rclcpp::get_logger("hello_moveit");
+
+    // Create a thread to run to spin a Executor
+    auto thread = std::thread([node]() {
+      rclcpp::executors::SingleThreadedExecutor executor;
+      executor.add_node(node);
+      executor.spin();
+    });
+
+    // Next step goes here
+
+    // Shutdown ROS and our thread
+    rclcpp::shutdown();
+    thread.join();
+    return 0;
+  }
+
+2.1 Build and Run
+~~~~~~~~~~~~~~~~~
+
+We will build and run the program to see that everything is right before we move on.
+
+Change directory back to the workspace directory ``ws_moveit2`` and run this command:
+
+.. code-block:: bash
+
+  colcon build --mixin debug
+
+After this succeeds, **open a new terminal**, then source the workspace environment script in that new terminal so that we can run our program.
+
+.. code-block:: bash
+
+  source install/setup.bash
+
+Run your program and see the output.
+
+.. code-block:: bash
+
+  ros2 run hello_moveit hello_moveit
+
+You will probably see this error:
+
+.. code-block::
+
+  terminate called after throwing an instance of 'rclcpp::exceptions::RCLError'
+    what():  the given context is not valid, either rcl_init() was not called or rcl_shutdown() was called., at /tmp/binarydeb/ros-rolling-rcl-4.0.0/src/rcl/guard_condition.c:67
+  [ros2run]: Aborted
+
+This is because ``rclcpp::shutdown()`` was called before the Executor finished initializing.
+We'll fix that in the next section by adding some code that does something after creating the thread.
+
+2.2 Examine the code
+~~~~~~~~~~~~~~~~~~~~
+
+The headers included at the top are just some standard C++ headers and the header for ROS and MoveIt that we will use later.
+
+After that we have the normal call to initialize rclcpp and then we create our Node.
+
+.. code-block:: C++
+
+  auto const node = std::make_shared<rclcpp::Node>(
+    "hello_moveit",
+    rclcpp::NodeOptions().automatically_declare_parameters_from_overrides(true)
+  );
+
+The first argument is the string that ROS will use to make a unique node.
+The second is needed for MoveIt because of how we use ROS Parameters.
+
+Before we can start use our ROS Node we have to give it to a spinning executor.
+This is what will enable ROS to call callbacks to update our Robot State among other things.
```

Comment:
I was confused for a while on why a thread is needed, as generally I see ROS nodes to execute `.spin` in the main thread. Having some more explanation of why this is needed may dispel the doubts I had when I first looked. I ended up looking at the code of `MoveGroupInterface`, and I wrote the following to myself:

> `MoveGroupInterface` requires a node as an input parameter, which we declared above. This is stored as the instance variable `node_` within it, and it is used to retrieve ROS parameters and the robot model. This node we declare should be executed in its own executor on a background thread, as `MoveGroupInterface` also defines its own internal/private node, stored on the variable `pnode_`. Publishers and subscribers are declared on `pnode_` to interact with move_group. Generally, methods such as `MoveGroupInterface::plan` and `MoveGroupInterface::execute` are called on the main thread. These methods calls `pnode_->spin_some()` internally, which effectively means the main thread of this code is running `pnode_` as opposed to the node we defined above (which again is only useful for parameters and robot models).

I'm not sure if this is accurate, as I'm still relatively new to MoveIt. If it is reasonably accurate, you're welcome to steal/improve it for this. That said, it might be a lot of detail that may be better suited to be in the MoveGroupInterface docs.

Further, I have no idea why MoveGroupInterface is designed the way it is, and I couldn't find an explanation of this (private node) design either. It could be good to have this explained as well (although this would also be likely better in the MoveGroupInterface documentation).

Reply:
I originally thought this was necessary, but I just tested it and we don't actually need this thread.  I'm going to remove it because it makes the tutorial simpler and is easier to read.

Reply:
I just tried this too and I get this error:

```
[demo-1] [INFO] [1644181939.795973661] [move_group_interface]: Ready to take commands for planning group manipulator.
[demo-1] [INFO] [1644181939.800464837] [moveit_ros.current_state_monitor]: Listening to joint states on topic 'joint_states'
[demo-1] [INFO] [1644181940.800568161] [moveit_ros.current_state_monitor]: Didn't received robot state (joint angles) with recent timestamp within 1.000000 seconds.
[demo-1] Check clock synchronization if your are running ROS across multiple machines!
[demo-1] [ERROR] [1644181940.800616142] [move_group_interface]: Failed to fetch current robot state
```

Then for me it just quits... I didn't really investigate too heavily into it, but when I add the thread back in, it appears to work..

Reply:
Try it again.  Here is the full source for the file you can copy-paste or compare with yours:

```cpp
#include <memory>
#include <thread>

#include <rclcpp/rclcpp.hpp>
#include <moveit/move_group_interface/move_group_interface.h>

int main(int argc, char * argv[])
{
  // Initialize ROS and create the Node
  rclcpp::init(argc, argv);
  auto const node = std::make_shared<rclcpp::Node>(
    "hello_moveit",
    rclcpp::NodeOptions().automatically_declare_parameters_from_overrides(true)
  );

  // Create a ROS logger
  auto const logger = rclcpp::get_logger("hello_moveit");

  // Create the MoveIt MoveGroup Interface
  using moveit::planning_interface::MoveGroupInterface;
  auto move_group_interface = MoveGroupInterface(node, "panda_arm");

  // Set a target Pose
  auto const target_pose = []{
    geometry_msgs::msg::Pose msg;
    msg.orientation.w = 1.0;
    msg.position.x = 0.28;
    msg.position.y = -0.2;
    msg.position.z = 0.5;
    return msg;
  }();
  move_group_interface.setPoseTarget(target_pose);

  // Create a plan to that target pose
  auto const [success, plan] = [&move_group_interface]{
    moveit::planning_interface::MoveGroupInterface::Plan msg;
    auto const ok = static_cast<bool>(move_group_interface.plan(msg));
    return std::make_pair(ok, msg);
  }();

  // Execute the plan
  if(success) {
    move_group_interface.execute(plan);
  } else {
    RCLCPP_ERROR(logger, "Planing failed!");
  }

  // Shutdown ROS and our thread
  rclcpp::shutdown();
  return 0;
}
```

Reply:
Here is the output I get when I run it.  Remember you need to have the demo launch file fully loaded before running the hello_moveit program:

```
> ros2 run hello_moveit hello_moveit
[INFO] [1644182399.668580614] [moveit_rdf_loader.rdf_loader]: Loaded robot model in 0.211959 seconds
[INFO] [1644182399.668850895] [moveit_robot_model.robot_model]: Loading robot model 'panda'...
[WARN] [1644182399.688569413] [moveit_ros.robot_model_loader]: No kinematics plugins defined. Fill and load kinematics.yaml!
[INFO] [1644182399.694424427] [move_group_interface]: Ready to take commands for planning group panda_arm.
[INFO] [1644182399.694576748] [move_group_interface]: MoveGroup action client/server ready
[INFO] [1644182399.996552370] [move_group_interface]: Planning request accepted
[INFO] [1644182400.397855526] [move_group_interface]: Planning request complete!
[INFO] [1644182400.498041963] [move_group_interface]: time taken to generate plan: 0.383186 seconds
[INFO] [1644182400.799244253] [move_group_interface]: Execute request accepted
[INFO] [1644182407.008906716] [move_group_interface]: Execute request success!
```

Reply:
Somewhat interesting, the thread revealed a race-condition in moveit that cory found and fixed here: https://github.com/ros-planning/moveit2/pull/1050

So if you haven't pulled moveit recently you might be running into that bug.

Reply:
Hmm thanks for the hints. I tried updating moveit to the lastest commit but it doesn't seem to work either. My movegroupinterface node is definitely started after the move_group nodes (in fact, I don't really quit that launch file as it contains move_group and rviz, both of which seem like they should be long running). 

I'm not sure what's up then. It's possible that I'm doing something else wrong. My setup is on a custom robot not panda, but it shouldn't be fundamentally any different. I'll take a closer look at it in a bit either in the debugger or I'll try it on panda when I get a chance later.

Reply:
@shuhaowu in case you didn't know about it we are hosting a virtual event on Feb24 for the community that uses MoveIt.  You can find the details here: https://discourse.ros.org/t/2022-annual-community-meeting/23331  We would be really interested to hear about how you've been using and anything that would make it better for you.

Again, thank you for helping me with this tutorial.  I would really like to make it easier to learn to use moveit.

Reply:
Thanks for the invite! I took a look at the event. It looks mostly like talks and not networking/chatting about MoveIt? Is that accurate? I'm reasonably busy in the morning but may be able to show up in the "happy hours" section of the event. Sorry for the late reply as I've been quite busy.

Reply:
You have the right idea about the event. I just wanted to make sure you knew about it if you wanted to come.